src/schema/schema.rs

```diff
@@ -33,263 +33,500 @@ lazy_static! {
 ///    required.
 ///
 /// [1]: http://json-schema.org
-#[derive(PartialEq, Clone, Debug)]
-pub struct Schema {
-  pub type_: SchemaType
+pub trait Schema {
+  /// Used to get a nested schema at a certain point.
+  fn get(&self, mut pointer: Pointer) -> Option<&Schema>;
+
+  /// Validates a query that a user would like to make on the database by
+  /// comparing it to the schema. Mostly checks that all properties described
+  /// in the query are accessible according to the schema.
+  fn validate_query(&self, query: &Query) -> Result<(), Error>;
 }
 
 impl Schema {
-  /// Gets a nested schema at a certain point.
-  pub fn get(&self, mut pointer: Pointer) -> Option<Self> {
+  /// Create a schema which does not run any validations.
+  pub fn none() -> SchemaNone {
+    SchemaNone::new()
+  }
+
+  /// Create a schema which ensures the value is null and only null.
+  pub fn null() -> SchemaNull {
+    SchemaNull::new()
+  }
+
+  /// Create a schema which validates a boolean primitive type.
+  pub fn boolean() -> SchemaBoolean {
+    SchemaBoolean::new()
+  }
+
+  /// Create a schema which validates a number.
+  pub fn number() -> SchemaNumber {
+    SchemaNumber::new()
+  }
+
+  /// Create a schema which validates a string.
+  pub fn string() -> SchemaString {
+    SchemaString::new()
+  }
+
+  /// Create a schema which validates an array. It takes another schema to
+  /// validate all of the child properties.
+  pub fn array() -> SchemaArray {
+    SchemaArray::new()
+  }
+
+  /// Create a schema which validates an object.
+  pub fn object() -> SchemaObject {
+    SchemaObject::new()
+  }
+
+  /// Creates a schema which validates enumerated values.
+  pub fn enum_<V>(values: Vec<V>) -> SchemaEnum where V: Into<Value> {
```

Comment:
You're right. There were a lot of warnings for the `unimplemented!` stuff so I missed this one.